### PR TITLE
fix: Fix various tests caused by cases byte size validation not handled properly

### PR DESCRIPTION
**Pull request by rmccorm4**
#### What does the PR do?

- Recent byte size validation checks were failing the L0_http HttpTest.test_byte unit test which sends a raw binary tensor.
- Recent input byte size validation checks are segfaulting for the CUDA Shared Memory case because it assumes it is in CPU memory and dereferences an invalid pointer.
- Currently, L0_trt_shape_tensor and L0_trt_reformat_free tests are failing because TensorRT infer requests should not check byte size.

The raw binary tensor in the failing test case consists of two memory chunks by the time it gets validated:
1. The size of the bytes element
2. The bytes element

This PR makes some generic refactoring to try to handle any series of buffers, where each element is not guaranteed to fit within a single buffer, and each buffer is not guaranteed to contain a single element. Also it skips byte size validation if input memory type is GPU or input platform is TensorRT.

#### Caveats: 

1. GPU buffers are currently skipped, as they may need some special handling to be dereferenced/checked compared to CPU buffers in the existing code. This will likely cause the element count checks to fail as well.
2. TensorRT inputs are currently skipped, as format-free IO tensors require communication with the backend to determine the byte size.
3. A check is made that the 4-byte `byte_size` indicator for a given element is contained within a single buffer, and is not split across buffers. If a byte_size is split across buffers, then an error is returned. This was just easier to implement, open to suggestions if anyone has a slick solution. Technically the buffer APIs should support doing this, but I assume it is unlikely to be done. Example of this currently unsupported case: 
    - `buffer1=[<size1><element1><size2_partial>]`
    - `buffer2=[<size2_partial><element2>...]`
    - `<size2_partial>` is split across buffers, and is currently rejected.

#### Open Items

- [ ] Verify test plan passes in CI
- [ ] Decide what to do about GPU tensors - probably need to handle them
- [ ] Decide what to do about TensorRT, e.g. shape tensors and reformat-free I/O tensors.

#### Checklist
- [X] PR title reflects the change and is of format `<commit_type>: <Title>`
- [X] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [X] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [X] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [x] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Related PRs:
https://github.com/triton-inference-server/server/pull/7326

#### Where should the reviewer start?
N/A

#### Test plan:
- L0_http
- L0_input_validation
- L0_infer_cudashm
- L0_sequence_batcher_cudashm
- L0_backend_output_detail
- L0_request_cancellation
- L0_trt_reformat_free
- L0_trt_shape_tensors
- Nightly pipeline

CI Pipeline ID: 15635924

#### Background
None

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
None

### CODE REVIEW

File: src/infer_request.cc
Comment (yinggeh) [inline src/infer_request.cc:1192]: I don't think we should return immediately incase more checks are appended to this function.

Comment (rmccorm4) [reply src/infer_request.cc:1192]: Open to changes on the GPU portions if you've already been looking into it. The early return can be changed if new checks are added or if we add support for checking the GPU buffers. 

The early return here is temporary to avoid failing the total element count validation at the end if we skip elements in gpu buffers.

Comment (yinggeh) [reply src/infer_request.cc:1192]: My concern is someone started appending new sections without looking backwards. You can change to
```
bool skip_byte_size_check = (expected_byte_size == -1);

if (!skip_byte_size_check &&
    ((byte_size > INT_MAX) || (static_cast<int64_t>(byte_size) != expected_byte_size))) {
```

Comment (yinggeh) [inline src/infer_request.cc:1192]: I don't think we should return immediately incase more checks are appended to this function.
My concern is someone will add new lines without looking backwards. You should consider
```
bool skip_byte_size_check = (expected_byte_size == -1);

if (!skip_byte_size_check &&
    ((byte_size > INT_MAX) || (static_cast<int64_t>(byte_size) != expected_byte_size))) {
```

File: src/infer_request.cc
Comment (rmccorm4) [reply src/infer_request.cc:1307]: I can update the message, it's probably best-effort on a per backend basis rather than strictly validated.

Were you working on updating the GPU buffer validation in core? What was your latest status regarding that? Feel free to slack.

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc:1192]: Agree that you should not return early on "success". And further more, I think you should reorder the code here because `ValidateBytesInputs()` already does "total byte size validation":
```
if (byte_data_type) {
  RETURN_IF_ERROR(ValidateBytesInputs(...));
}
else { // non_byte_data
  // line 1179-1185
  // line 1195
  // line 1197-1205
}
```

File: src/infer_request.cc
Comment (GuanLuo) [inline src/infer_request.cc:1293]: Why not like below to keep variables in scope
```
for (buf : buffers) {
  buffer_size = ...;
  for (checked_size = 0; checked_size < buffer_size) {
    ..
  }
}
```

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc:1307]: Slacked you.

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc]: I also just realized that all these checks are inside of a for loop, so you probably intended to`continue` instead of `return` (neither was a good idea). It's prone to make mistakes which makes me feel like to refactor such a long function `Normalize`.

Comment (yinggeh) [reply src/infer_request.cc]: > Agree that you should not return early on "success". And further more, I think you should reorder the code here because `ValidateBytesInputs()` already does "total byte size validation":
> 
> ```
> if (byte_data_type) {
>   RETURN_IF_ERROR(ValidateBytesInputs(...));
> }
> else { // non_byte_data
>   // line 1179-1185
>   // line 1195
>   // line 1197-1205
> }
> ```

`ValidateBytesInputs` asserts all buffers and all strings are visited. It doesn't check the size of the last string imo. I still think it's a good idea to compare expected_byte_size at one place regardless of input data type.

File: src/infer_request.cc
Comment (yinggeh) [inline src/infer_request.cc:1335]: Please revisit the logic in this case. I think it will cause segfault/infinite loop if `remaining_buffer_size == remaining_element_size`.

**Comment by rmccorm4** [comment]
Re-assigning to @yinggeh as this involves a lot of refactoring to the original validation logic feature he wrote and it seems like he has some ideas for further refactoring.

File: src/infer_request.cc
Comment (rmccorm4) [inline src/infer_request.cc:1209]: This was my mistake, but let's rename `byte_size_valid` to `byte_size_invalid` or similar.

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc]: Let me know if there is misalignment in my understanding:

"total byte size" in this function scope is `byte_size` which is set to `input.Data()->TotalByteSize()`.

In `ValidateBytesInputs()`, it iterates all elements in the data to get the actual byte size used in execution, and compare it against the buffer boundary (using buffer count and buffer byte size). Isn't the buffer boundary the reflection of `TotalByteSize()`? If iteration exactly ends at buffer boundary, then it implies the `expected size == got size`

File: src/infer_request.cc
Comment (GuanLuo) [inline src/infer_request.cc:1314]: Isn't this already checked by the while loop conditional? The largest value of `buffer_next_idx` at the beginning of the loop will be `buffer_count-1`, after line 1310 will be `buffer_count` and this line will never be true.

File: src/infer_request.cc
Comment (rmccorm4) [reply src/infer_request.cc:1189]: Yes this is a WAR to avoid breaking existing TRT reformat_free features. Shape tensors can be explicitly checked, but reformat_free is not as easy to check without some communication with the backend or something. We wanted to unblock Sai who is working on other TRT features and avoid noise in failing TRT tests until we have a proper fix for this.

Comment (GuanLuo) [reply src/infer_request.cc:1189]: Can you or Yingge add some tracking info here if it will be addressed later?

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc:1314]: `buffer_next_idx > buffer_count` statement is true if `remaining_buffer_size>0` and `buffer_next_idx == buffer_count` at the top while loop condition check. This happens when the last string element content is in the last two buffers but the second buffer somehow does not exist.

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc:1189]: Adding TODOs?

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc]: If iteration exactly ends at buffer boundary, it only means the number of bytes matches `kElementSizeIndicator`.

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc:1314]: >This happens when the last string element content is in the last two buffers but the second buffer somehow does not exist.

Can you point me to the test case that illustrate this edge case? I can't see why the 2nd buffer will not exist while input data says it contains 2 buffer.

> if `remaining_buffer_size>0` and `buffer_next_idx == buffer_count`
if `remaining_buffer_size>0`, then line 1307 `if (!remaining_buffer_size)` is false and won't hit this line.

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc]: "Iteration ends" refer to the point after checking all the elements, does that make sense?

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc:1189]: Add ticket number here and add comment in the ticket itself to revert / address this change as part of definition of done

**Review by GuanLuo — COMMENTED** [review]
See new comments in opened discussion.

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc:1314]: You are right. This check is redundent.

File: src/infer_request.cc
Comment (yinggeh) [reply src/infer_request.cc]: Gotcha. Here `byte_size` is comparing to itself (expected_byte_size) which is redundant.

Comment (yinggeh) [reply src/infer_request.cc]: Dumb question: Is `input.Data()->TotalByteSize()` guarenteed to be accurate? I can see this check is useful when there is a size mismatch between `byte_size` and bytes get read from all buffers.

**Comment by yinggeh** [comment]
> Please update the description with the latest information and pipeline ID, I need to see the test status

I believe I have updated with the latest info and pipeline ID. Is there anything particular you think is missing?

File: src/infer_request.cc
Comment (rmccorm4) [reply src/infer_request.cc:1190]: Looks like ticket numbers, details, and follow-ups were added - resolving

**Review by rmccorm4 — COMMENTED** [review]
I can't approve this PR because I originally opened it, but I am verbally approving this if all the affected CI tests in the test plan description are passing.

I added a follow-up ticket DLIS-6833 to address some of the logical refactoring brought up by Guan after code freeze.

**Review by GuanLuo — APPROVED** [review]

File: src/infer_request.cc
Comment (GuanLuo) [reply src/infer_request.cc]: >Is input.Data()->TotalByteSize() guaranteed to be correct?

This function should return the total size of the buffers.

> mismatch between `byte_size` and bytes get read from all buffers

From the current implementation they are the same thing, right? I think the comparison should be between buffer byte size (from `TotalByteSize()`) and byte size computed from other sources (i.e. shape * data byte size for non-byte tensor, and iterating elements for byte tensor).

**Comment by yinggeh** [comment]
Please also review related server PR github.com/triton-inference-server/server/pull/7326

**Pull request merged**
